### PR TITLE
Create a tool to validate a staging area (#2742)

### PR DESCRIPTION
**Pull request by danielsotirhos**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented on demo expectations                        <sub>or labelled PR as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that Demo expectations are clear              <sub>or PR is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and build passes in `sandbox` <sub>or added `no sandbox` label</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Purchased BQ slot committment in `dev`                <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Deleted BQ slot committment in `dev`                  <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing `dev`</sub>
- [ ] Purchased BQ slot committment in `prod`               <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Deleted BQ slot committment in `prod`                 <sub>or this PR does not require reindexing `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

File: scripts/validate_staging_area.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+                      len(self.file_errors))
+        return exit_code
+
+    @classmethod
+    def _parse_args(cls, argv):
+        parser = argparse.ArgumentParser(description=__doc__)
+        parser.add_argument('--staging-area', '-s',
+                            required=True,
+                            help='The Google Cloud Storage URL of the staging area. '
+                                 'Syntax is gs://<bucket>[/<path>].')
```
Comment (achave11) [inline scripts/validate_staging_area.py:70]: I think this might qualify for wrapping all elements.

File: scripts/validate_staging_area.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+    def download_blob_as_json(self, blob: gcs.Blob) -> Optional[JSON]:
+        file_json = json.loads(blob.download_as_string())
+        return file_json
+
+    def validate_links_file(self, blob: gcs.Blob) -> None:
+        # Expected syntax: links/{bundle_uuid}_{version}_{project_uuid}.json
+        _, _, file_name = blob.name.rpartition('/')
+        assert file_name.count('_') == 2
+        assert file_name.endswith('.json')
+        bundle_uuid, _, project_uuid = file_name[:-5].split('_')
```
Comment (achave11) [inline scripts/validate_staging_area.py:142]: Is it necessary to extract the `bundle_uuid` although it isn't being used?

**Review by achave11 — COMMENTED** [review]
Minor nits/questions, but I leave it at your discretion.

File: scripts/validate_staging_area.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielsotirhos) [reply scripts/validate_staging_area.py:70]: I think the long and short forms of the argument reads better when on the same line, and it matches the use of `add_argument` elsewhere in our codebase. I agree however that the other parameters should be wrapped.

[the remark this answers is not in the document]
Comment (hannes-ucsc) [reply scripts/validate_staging_area.py:70]: I agree.

**Review by achave11 — APPROVED** [review]
Great work, approved!

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
We'll demo this as part of the DCP/-wide Tuesday demo.

**Pull request merged**